src/ai/defaultai_seafaring.cc

```diff
@@ -546,6 +606,83 @@ void DefaultAI::expedition_management(ShipObserver& so) {
 	so.waiting_for_command_ = false;
 }
 
+// this is called whenever a warship received a notification that requires
+// navigation decisions (these notifications are processes not in 'real time')
+void DefaultAI::warship_management(ShipObserver& so) {
+	const Time& gametime = game().get_gametime();
+
+	game().send_player_warship_command(*so.ship, Widelands::WarshipCommand::kSetCapacity, {0u});
+
+	if (!so.guarding) {
+
+		if (tradeship_refit_needed) {
+			// Let's keep this simple. (for now?)
+			// We only backfit if we find an idle warship. Trade ships shouldn't disappear anyway, so
+			// backfit requests should be extremely rare.
+
+			verb_log_dbg_time(gametime, "AI %d: Refit warship %s to trade ship",
+			                  player_->player_number(), so.ship->get_shipname().c_str());
+			game().send_player_refit_ship(*so.ship, Widelands::ShipType::kTransport);
+			tradeship_refit_needed = false;
+			--warships_count;
+			so.last_command_time = gametime;
+			so.waiting_for_command_ = false;
+			return;
+		}
+
+		verb_log_dbg_time(gametime, "AI %d: Warship %s is looking for assignment.",
+		                  player_->player_number(), so.ship->get_shipname().c_str());
+		PortSiteObserver* picked_port = nullptr;
+		Widelands::Quantity picked_port_guard_ships = kWarshipsPerPort + 1;
+
+		for (PortSiteObserver& p_obs : portsites) {
+			verb_log_dbg_time(gametime, "AI %d: Port %s has %d ships assigned",
+			                  player_->player_number(), p_obs.site->get_warehouse_name().c_str(),
+			                  p_obs.ships_assigned);
+			if (p_obs.ships_assigned < picked_port_guard_ships) {
+				picked_port = &p_obs;
+				picked_port_guard_ships = p_obs.ships_assigned;
+				if (picked_port_guard_ships == 0) {
+					break;
+				}
+			}
+		}
+
+		if (picked_port != nullptr && picked_port_guard_ships < kWarshipsPerPort) {
+			assert(picked_port_guard_ships <= kWarshipsPerPort);
+			assert(picked_port_guard_ships == picked_port->ships_assigned);
+
+			verb_log_dbg_time(gametime, "AI %d: Assigning warship %s to port %s",
+			                  player_->player_number(), so.ship->get_shipname().c_str(),
+			                  picked_port->site->get_warehouse_name().c_str());
+			game().send_player_ship_set_destination(*so.ship, picked_port->site->get_portdock());
+			++(picked_port->ships_assigned);
+			so.guarding = true;
+		} else {
+#ifndef NDEBUG
+			if (picked_port == nullptr) {
+				assert(portsites.empty());
+			} else {
+				assert(picked_port_guard_ships == kWarshipsPerPort);
+				assert(picked_port_guard_ships == picked_port->ships_assigned);
+			}
+#endif
+			log_dbg_time(gametime, "AI %d: all ports have enough guard ships, warship %s remains idle",
```

Comment:
I always thought about how to increase the max number per port. maybe we could just increase this by 1 if this is the case.

Reply:
That would mess up the refit needed calculations.

But I've just thought of a quick and dirty workaround...